### ISSUE TITLE
Bug: [@typescript-eslint/no-invalid-void-type] allowInGenericTypeArguments not working as documented

### ISSUE THREAD

**Issue by trajano** [typescript-eslint/typescript-eslint#6340]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [X] I have tried restarting my IDE and the issue persists.
- [X] I have updated to the latest version of the packages.
- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22package%3A+eslint-plugin%22) and found none that matched my issue.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Playground Link

https://typescript-eslint.io/play/#ts=4.8.4&sourceType=module&code=GYVwdgxgLglg9mABAJwKZgCauQCTnAawB4AVABWTgAcBnAGkRICVUaQAbKAPgAoIBDduwBG-CAQBciHlUq0p5OTQCUiALxdGLNpwZUFFaisQBvALAAoRIgFDR4mcoDclgL6XQkWAkQgaqADlUVAwaFmA0GgALAFFgYFRoHlVzCzcPcGh4JH8oAEEQKCiAZSh+KFRk00t0izRMbDxCIktrGihkGDAAczpWxAA3OBgMS14ZJVUNX38gkLDUCNZY+MSoZIYAIkF2OE3nIA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1uYDcBDRgBNaPDpWFFS6MAG1w2HAHc+0JpAA08hVgXZIA+B0UBJJgHFETRFTIAVEogCC0AOawAtpfyoM+OIi1sAF9AgF15EKCgA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA

(note it does not seem to reproduce the error there)

### Repro Code

```typescript
const { rerender: rerenderNeedsRefresh } = renderHook<
    NeedsRefreshEffectProps,
    void
  >((props) => useNeedsRefreshEffect(props), {
    initialProps: {
      authState: AuthState.INITIAL,
      setAuthState,
      backendReachable: true,
      onRefreshError,
      refreshAsync,
      notify,
    },
  });
```

### ESLint Config

```javascript
"eslintConfig": {
    "root": true,
    "extends": [
      "@react-native-community",
      "plugin:@typescript-eslint/strict",
      "plugin:jest/recommended",
      "plugin:react-hooks/recommended",
      "plugin:testing-library/react",
      "prettier"
    ],
    "parserOptions": {
      "project": [
        "./tsconfig.json"
      ]
    }
  },

```

### Expected Result

no warnings

### Actual Result

warning  void is only valid as a return type or generic type argument  @typescript-eslint/no-invalid-void-type

**Comment by bradzacher** [comment]
Your playground link doesn't have any errors. Could you please update it to repro your issue?

**Comment by trajano** [comment]
I noticed that.  Not sure why on my local development environment it's showing.  I did `yarn outdated` as well and didn't see anything odd.

**Comment by trajano** [comment]
The documentation states that 

```
logAndReturn<void>(undefined);
```

is incorrect, but https://typescript-eslint.io/rules/no-invalid-void-type/#allowingenerictypearguments states

> This option is true by default.

So the sample should work by default

**Comment by bradzacher** [comment]
The playground does not support recommended configs. This is on purpose because we want ***minimal*** bug reports.
Turning on an entire config is not minimal at all!

**Comment by trajano** [comment]
ok updated it now shows the error switching the allowInGenericTypes from true to false does not change the error output

**Comment by bradzacher** [comment]
This is actually working as intended, as the `allowInGenericTypeArguments` option specifically only applies to generic type parameters for type references (eg `type T = Promise<void>`).

This isn't explicitly called out, but is implicitly shown due to the fact that `logAndReturn<void>(undefined);` is never listed in a valid example anywhere.

Duplicate of https://github.com/typescript-eslint/typescript-eslint/issues/3612
